GeeksCoreLibrary/Components/Account/Services/AccountsService.cs

```diff
@@ -447,5 +454,199 @@ public async Task<List<RoleModel>> GetUserRolesAsync(ulong userId, bool includeP
+                return;
+            }
+
+            if (userIdForGoogleCid <= 0)
+            {
+                var sanitizedGoogleClientIdCookieValue = googleClientIdCookieValue.Replace(Environment.NewLine, "").Replace("\n", "").Replace("\r", "");
+                logger.LogTrace($"Google Analytics cookie found ({Constants.GoogleAnalyticsCookieName}: {sanitizedGoogleClientIdCookieValue}), but the user is not logged in and we did not create a new account (successfully).", true);
+                return;
+            }
+
+            var googleClientId = String.Join(".", clientIdSplit.Skip(2));
+            
+            
+            var wiserItemsService = ActivatorUtilities.CreateInstance<WiserItemsService>(serviceProvider, databaseConnection);
+            var tablePrefix = await wiserItemsService.GetTablePrefixForEntityAsync(settings.EntityType);
+
+            var detail = new WiserItemDetailModel()
+            {
+                Key = String.IsNullOrWhiteSpace(settings.GoogleClientIdFieldName) ? Constants.DefaultGoogleCidFieldName : settings.GoogleClientIdFieldName,
+                Value = googleClientId
+            };
+
+            await wiserItemsService.SaveItemDetailAsync(detail, userIdForGoogleCid, entityType: settings.EntityType);
+        }
+
+        /// <summary>
+        /// Saves a login attempt in the details of the user. This will use the query in <see cref="Settings.SaveLoginAttemptQuery"/>.
+        /// </summary>
+        /// <param name="success">Whether the login attempt was successful or not.</param>
+        /// <param name="userId">The ID of the user that is attempting to login.</param>
+        /// <param name="extraDataForReplacements"></param>
+        /// <param name="settings"></param>
+        public async Task SaveLoginAttemptAsync(bool success, ulong userId,  Dictionary<string,string> extraDataForReplacements, AccountCmsSettingsModel settings)
+        {
+            var query = SetupAccountQuery(settings.SaveLoginAttemptQuery, settings, userId, success: success);
+            if (String.IsNullOrWhiteSpace(query))
+            {
+                return;
+            }
+
+            await componentsService.RenderAndExecuteQueryAsync(query, extraDataForReplacements, skipCache: true);
+        }
+
+        /// <summary>
+        /// Do all default login replacements on a SQL template and adds the variables to the <see cproperty="SystemConnection"/>.
+        /// </summary>
+        /// <param name="template"></param>
+        /// <param name="settings"></param>
+        /// <param name="userId"></param>
+        /// <param name="loginValue"></param>
+        /// <param name="emailAddress"></param>
+        /// <param name="token"></param>
+        /// <param name="success"></param>
+        /// <param name="passwordHash"></param>
+        /// <param name="subAccountId"></param>
+        /// <param name="role"></param>
+        /// <returns></returns>
```

Comment:
The summary can be replaced with `/// <inheritdoc />`, then the summary of the interface is shown so we only need to manage this on one place.